## Grouping Data & Aggregating Variables

### 1. Import Libraries and Dataframe

In [6]:
#Importing Libraries
import pandas as pd
import numpy as ny
import os

In [8]:
#Creating Path
path = r'C:\Users\ankit\Desktop\DataAnalysisCF\DataImmersion\PythonFundamentals\Instacart Basket Analysis\02 Data'

In [10]:
#Importing Dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path,'Prepared Data','orders_products_merged_derived.pkl'))

### 2. Find the aggregated mean of the “order_number” column grouped by “department_id” for entire dataframe. 

In [15]:
# Use groupby and agg functions to calculate mean of order_number

ords_prods_merge.groupby('department_id')['order_number'].mean()


department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

### 3.Analyze the result.

Results from entire dataframe include all the department ids, whereas subset has only few department_ids. Department_id 21 has the highest mean of 22.9 and department_id 5 has lowest mean of 15.2. There is change in the mean value for all departments , they have mostly reduced when entire dataframe is condidered.

### 4. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [24]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(ny.max)


C:\Users\ankit\AppData\Local\Temp\ipykernel_21744\2281651034.py:1: FutureWarning: The provided callable <function max at 0x0000026A178A6160> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby('user_id')['order_number'].transform(ny.max)


In [26]:
ords_prods_merge.head(50)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   
20           1  Chocolate Sandwich Cookies        61             19     5.8   
21           1  Chocolate Sandwich Cookies        61             19     5.8   
22           1  Chocolate Sandwich Cookies        61             19     5.8   
23           1  Chocolate Sandwich Cookies        61             19     5.8   
24           1  Chocolate Sandwich Cookies        61             19     5.8   
25           1  Chocolate Sandwich Cookies        61             19     5.8   
26           1  Chocolate Sandwich Cookies        61             19     5.8   
27           1  Chocolate Sandwich Cookies        61             19     5.8   
28           1  Chocolate Sandwich Cookies        61             19     5.8   
29           1  Chocolate Sandwich Cookies        61             19     5.8   
30           1  Chocolate Sandwich Cookies        61             19     5.8   
31           1  Chocolate Sandwich Cookies        61             19     5.8   
32           1  Chocolate Sandwich Cookies        61             19     5.8   
33           1  Chocolate Sandwich Cookies        61             19     5.8   
34           1  Chocolate Sandwich Cookies        61             19     5.8   
35           1  Chocolate Sandwich Cookies        61             19     5.8   
36           1  Chocolate Sandwich Cookies        61             19     5.8   
37           1  Chocolate Sandwich Cookies        61             19     5.8   
38           1  Chocolate Sandwich Cookies        61             19     5.8   
39           1  Chocolate Sandwich Cookies        61             19     5.8   
40           1  Chocolate Sandwich Cookies        61             19     5.8   
41           1  Chocolate Sandwich Cookies        61             19     5.8   
42           1  Chocolate Sandwich Cookies        61             19     5.8   
43           1  Chocolate Sandwich Cookies        61             19     5.8   
44           1  Chocolate Sandwich Cookies        61             19     5.8   
45           1  Chocolate Sandwich Cookies        61             19     5.8   
46           1  Chocolate Sandwich Cookies        61             19     5.8   
47           1  Chocolate Sandwich Cookies        61             19     5.8   
48           1  Chocolate Sandwich Cookies        61             19     5.8   
49           1  Chocolate Sandwich Cookies        61  

### 4. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [106]:
ords_prods_merge.loc[ords_prods_merge['max_order']> 40, 'loyalty_flag'] = 'Loyal Customer'

In [30]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

In [32]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10 , 'loyalty_flag'] = 'New Customer'

In [34]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

   first_order  add_to_cart_order  reordered  _merge    price_range_loc  \
0        False                  5          0    both  Mid-range product   
1        False                  1          1    both  Mid-range product   
2        False                 20          0    both  Mid-range product   
3         True                 10          0    both  Mid-range product   
4        False                 11          1    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly Busy  Regularly Busy Days           Most orders        32   
1  Regularly Busy  Regularly Busy Days        Average orders        32   
2     Busiest Day         Busiest Days        Average orders         5   
3  Regularly Busy      Least Busy Days           Most orders         3   
4      Least Busy      Least Busy Days        Average orders         3   

       loyalty_flag  
0  Regular Customer  
1  Regular Customer  
2      New Customer  
3      New Customer  
4      New Customer  

[5 rows x 21 columns]

In [36]:
# Check frequency of values in new loyalty flag column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

### 5. Look at Spending Habits of Loyalty Groups

The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [40]:
# Use groupby and agg() to determine basic statisitcs of spending habits by loyalty group
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})


prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer    10.386336  1.0  99999.0
New Customer      13.294670  1.0  99999.0
Regular Customer  12.495717  1.0  99999.0

The mean value of prices of Loyal customer is lower than New Customer by about 3$, which means the prices of products purchased by Loyal customers is much lower than New Customers.

### 6.

The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”


In [49]:
# Creating a new column mean_price
ords_prods_merge['mean_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(ny.mean)

C:\Users\ankit\AppData\Local\Temp\ipykernel_21744\3325414555.py:2: FutureWarning: The provided callable <function mean at 0x0000026A178A6B60> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(ny.mean)


In [50]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

   add_to_cart_order  reordered  _merge    price_range_loc     busiest_day  \
0                  5          0    both  Mid-range product  Regularly Busy   
1                  1          1    both  Mid-range product  Regularly Busy   
2                 20          0    both  Mid-range product     Busiest Day   
3                 10          0    both  Mid-range product  Regularly Busy   
4                 11          1    both  Mid-range product      Least Busy   

          busiest_days busiest_period_of_day max_order      loyalty_flag  \
0  Regularly Busy Days           Most orders        32  Regular Customer   
1  Regularly Busy Days        Average orders        32  Regular Customer   
2         Busiest Days        Average orders         5      New Customer   
3      Least Busy Days           Most orders         3      New Customer   
4      Least Busy Days        Average orders         3      New Customer   

   mean_price  
0    6.935811  
1    6.935811  
2    7.930208  
3    4.972414  
4    4.972414  

[5 rows x 22 columns]

In [53]:
#Creating new column spender_type based on average expenditure
ords_prods_merge.loc[ords_prods_merge['mean_price']< 10 , 'spender_type'] = 'Low Spender'

In [55]:
ords_prods_merge.loc[ords_prods_merge['mean_price']>= 10 , 'spender_type'] = 'High Spender'

In [67]:
#Checking frequency
ords_prods_merge['spender_type'].value_counts(dropna = False)

spender_type
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

In [63]:
#Checking relavent columns
ords_prods_merge[['user_id','mean_price','spender_type']].head(25)

user_id  mean_price spender_type
0       138    6.935811  Low Spender
1       138    6.935811  Low Spender
2       709    7.930208  Low Spender
3       764    4.972414  Low Spender
4       764    4.972414  Low Spender
5       777    6.935398  Low Spender
6       825    5.957576  Low Spender
7       910    6.680000  Low Spender
8      1052    7.162500  Low Spender
9      1052    7.162500  Low Spender
10     1379    5.813636  Low Spender
11     1480    5.081633  Low Spender
12     1480    5.081633  Low Spender
13     1480    5.081633  Low Spender
14     1494    6.803797  Low Spender
15     1494    6.803797  Low Spender
16     1494    6.803797  Low Spender
17     1540    6.607122  Low Spender
18     1540    6.607122  Low Spender
19     1540    6.607122  Low Spender
20     1540    6.607122  Low Spender
21     1540    6.607122  Low Spender
22     1540    6.607122  Low Spender
23     1540    6.607122  Low Spender
24     1540    6.607122  Low Spender

### 7. Create frequency flag to determine frequent and non frequent flag

In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [72]:
#Creating new column median_prior_order
ords_prods_merge['median_prior_order'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(ny.median)

C:\Users\ankit\AppData\Local\Temp\ipykernel_21744\3889968745.py:2: FutureWarning: The provided callable <function median at 0x0000026A179D58A0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_prior_order'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform(ny.median)


In [74]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0   3139998      138            28                  6                 11  ...   
1   1977647      138            30                  6                 17  ...   
2    389851      709             2                  0                 21  ...   
3    652770      764             1                  3                 13  ...   
4   1813452      764             3                  4                 17  ...   

   _merge    price_range_loc     busiest_day         busiest_days  \
0    both  Mid-range product  Regularly Busy  Regularly Busy Days   
1    both  Mid-range product  Regularly Busy  Regularly Busy Days   
2    both  Mid-range product     Busiest Day         Busiest Days   
3    both  Mid-range product  Regularly Busy      Least Busy Days   
4    both  Mid-range product      Least Busy      Least Busy Days   

  busiest_period_of_day max_order      loyalty_flag mean_price spender_type  \
0           Most orders        32  Regular Customer   6.935811  Low Spender   
1        Average orders        32  Regular Customer   6.935811  Low Spender   
2        Average orders         5      New Customer   7.930208  Low Spender   
3           Most orders         3      New Customer   4.972414  Low Spender   
4        Average orders         3      New Customer   4.972414  Low Spender   

   median_prior_order  
0                 8.0  
1                 8.0  
2                 8.0  
3                 9.0  
4                 9.0  

[5 rows x 24 columns]

In [76]:
# Create order frequency flag based on given criteria using loc function
ords_prods_merge.loc[ords_prods_merge['median_prior_order']> 20 , 'frequency'] = 'Non-frequent customer'

In [86]:
ords_prods_merge.loc[(ords_prods_merge['median_prior_order']> 10) & (ords_prods_merge['median_prior_order']<= 20) , 'frequency'] = 'Regular customer'

In [80]:
ords_prods_merge.loc[ords_prods_merge['median_prior_order'] <= 20 , 'frequency'] = 'Frequent customer'

In [99]:
#Checking freqency of the new column
ords_prods_merge['frequency'].value_counts(dropna = False)

frequency
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

There are 5 values for NAN , which may be due to bank values in 'days since prior order'. As it is first order the values are blank. 

In [97]:
# Checking relevant columns
ords_prods_merge[['user_id','median_prior_order','frequency']].head(25)

user_id  median_prior_order          frequency
0       138                 8.0  Frequent customer
1       138                 8.0  Frequent customer
2       709                 8.0  Frequent customer
3       764                 9.0  Frequent customer
4       764                 9.0  Frequent customer
5       777                11.0   Regular customer
6       825                20.0   Regular customer
7       910                 6.0  Frequent customer
8      1052                10.0  Frequent customer
9      1052                10.0  Frequent customer
10     1379                20.0   Regular customer
11     1480                 7.0  Frequent customer
12     1480                 7.0  Frequent customer
13     1480                 7.0  Frequent customer
14     1494                 7.0  Frequent customer
15     1494                 7.0  Frequent customer
16     1494                 7.0  Frequent customer
17     1540                 7.0  Frequent customer
18     1540                 7.0  Frequent customer
19     1540                 7.0  Frequent customer
20     1540                 7.0  Frequent customer
21     1540                 7.0  Frequent customer
22     1540                 7.0  Frequent customer
23     1540                 7.0  Frequent customer
24     1540                 7.0  Frequent customer

### 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [111]:
ords_prods_merge.to_pickle(os.path.join(path, 'Prepared Data', 'orders_products_merged_derived_grouped.pkl'))